staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,339 @@
+		return
+	}
+
+	// otherwise, randomly select an apiserver
+	rand := rand.Intn(len(serviceableByResp.serviceableBy))
+	apiserverId := serviceableByResp.serviceableBy[rand]
+
+	// finally proxy
+	ip, err := reconciler.GetEndpoint(apiserverId)
+	if err != nil {
+		klog.ErrorS(err, "error proxying request for the requested GVR")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable(fmt.Sprintf("Error proxying request for the requested GVR: %v, err: %v", gvr, err)), s, gv, w, req)
+	}
+
+	// TODO: currently only looking at the first element of ips, check how to handle other values
+	if ip == "" {
+		klog.Errorf("Failed to serve request: Found no endpoints in server leases for the remote apiserver")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Found no endpoints in servers leases for the remote apiserver"), s, gv, w, req)
+		return
+	}
+
+	if len(strings.Split(ip, ":")) != 2 {
+		klog.ErrorS(err, "error getting ip and port info of the remote server while proxying")
+		responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), s, gv, w, req)
+		return
+	}
+
+	h.proxyRequestToDestinationAPIServer(req, w, ip)
+
+}
+
+func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string) (serviceableByResponse, error) {
+
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+	apiserversi, ok := h.svMap.Load(gvr)
+
+	// no value found for the requested gvr in svMap
+	if !ok || apiserversi == nil {
+		return serviceableByResponse{}, fmt.Errorf("no StorageVersions found for the GVR: %v", gvr)
+	}
+	apiservers := apiserversi.(*sync.Map)
+	response := serviceableByResponse{}
+	var serviceableBy []string
+	apiservers.Range(func(key, value interface{}) bool {
+		apiserverKey := key.(string)
+		if apiserverKey == localAPIServerId {
+			response.locallyServiceable = true
+			return false
+		}
+		serviceableBy = append(serviceableBy, apiserverKey)
+		return true
+	})
+	response.serviceableBy = serviceableBy
+	return response, nil
+}
+
+func (h *peerProxyHandler) proxyRequestToDestinationAPIServer(req *http.Request, w http.ResponseWriter, hostname string) {
+	user, ok := apirequest.UserFrom(req.Context())
+	if !ok {
+		klog.Errorf("failed to get user info from request")
+		return
+	}
+
+	// write a new location based on the existing request pointed at the target service
+	location := &url.URL{}
+	location.Scheme = "https"
+	location.Host = hostname
+	location.Path = req.URL.Path
+	location.RawQuery = req.URL.Query().Encode()
+
+	newReq, cancelFn := apiserverproxyutil.NewRequestForProxy(location, req)
+	newReq.Header.Add(PeerProxiedHeader, "true")
+	defer cancelFn()
+
+	// create transport
+	clientConfig := &transport.Config{
+		TLS: transport.TLSConfig{
+			Insecure:   false,
+			CertFile:   h.proxyClientCertFile,
+			KeyFile:    h.proxyClientKeyFile,
+			CAFile:     h.peerCAFile,
+			ServerName: "kubernetes.default.svc",
```

Comment:
is this guarantee?

Reply:
This is what we decided https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy#proxy-transport-between-apiservers-and-authn

Unsure if I need to do some validation in the code for this.